mlflow/utils/virtualenv.py

```diff
@@ -0,0 +1,254 @@
+
+
+def _get_python_env(local_model_path):
+    """
+    Constructs `PythonEnv` from the model artifacts stored in `local_model_path`. If
+    `python_env.yaml` is available, use it, otherwise extract model dependencies from `conda.yaml`.
+    If `conda.yaml` contains conda dependencies except `python`, `pip`, `setuptools`, and, `wheel`,
+    an `MlflowException` is thrown because conda dependencies cannot be installed in a virtualenv
+    environment.
+
+    :param local_model_path: Local directory containing the model artifacts.
+    :return: `PythonEnv` instance.
+    """
+    python_env_file = local_model_path / _PYTHON_ENV_FILE_NAME
+    requirements_file = local_model_path / _REQUIREMENTS_FILE_NAME
+    conda_env_file = local_model_path / _CONDA_ENV_FILE_NAME
+    if python_env_file.exists():
+        return PythonEnv.from_yaml(python_env_file)
+    else:
+        _logger.info(
+            "This model is missing %s, which is because it was logged in an older version"
+            "of MLflow (< 1.26.0) that does not support restoring a model environment with "
+            "virtualenv. Attempting to extract model dependencies from %s and %s instead.",
+            _PYTHON_ENV_FILE_NAME,
+            _REQUIREMENTS_FILE_NAME,
+            _CONDA_ENV_FILE_NAME,
+        )
+        conda_deps = _get_conda_dependencies(conda_env_file)
+        build_packages = ("python", *PythonEnv.BUILD_PACKAGES)
+        conda_deps = [d for d in conda_deps if _get_package_name(d) not in build_packages]
+        if conda_deps:
+            raise MlflowException(
+                f"Cannot restore this model's environment with virtualenv because it contains "
+                f"conda dependencies: {conda_deps}."
+            )
+        if requirements_file.exists():
+            deps = PythonEnv.get_dependencies_from_conda_yaml(conda_env_file)
+            deps.pop("dependencies", None)
+            return PythonEnv(
+                **deps,
+                dependencies=[f"-r {_REQUIREMENTS_FILE_NAME}"],
+            )
+        else:
+            return PythonEnv.from_conda_yaml(conda_env_file)
+
+
+def _create_virtualenv(local_model_path, python_bin_path, env_dir, python_env):
+    # Created a command to activate the environment
+    paths = ("bin", "activate") if _IS_UNIX else ("Scripts", "activate.bat")
+    activate_cmd = env_dir.joinpath(*paths)
+    activate_cmd = f"source {activate_cmd}" if _IS_UNIX else activate_cmd
+
+    if env_dir.exists():
+        _logger.info("Environment %s already exists", env_dir)
+        return activate_cmd
+
+    _logger.info("Creating a new environment %s", env_dir)
+    _exec_cmd(["virtualenv", "--python", python_bin_path, env_dir], capture_output=False)
+
+    _logger.info("Installing dependencies")
+    for deps in filter(None, [python_env.build_dependencies, python_env.dependencies]):
+        # Use a unique name to avoid conflicting custom requirements files logged by a user
```

Comment:
Is there a reason we have to place the requirements.txt file in the same directory as the model? If not, can we write it to a different temp directory?

Reply:
Good point, I updated the implementation to avoid modifying the model contents. Could you take a look?